core/src/main/clojure/xtdb/expression/list.clj

```diff
@@ -0,0 +1,26 @@
+(ns xtdb.expression.list
+  (:require [xtdb.expression :as expr]
+            [xtdb.types :as types]
+            [xtdb.util :as util])
+  (:import (xtdb.arrow ListExpression RelationReader ListValueReader)))
+
+(def vec-writer-sym (gensym 'vec_writer))
+
+(def compile-list-expr
+  (-> (fn [expr opts]
+        (let [expr (expr/prepare-expr expr)
+              {:keys [return-type continue] :as emitted-expr} (expr/codegen-expr expr opts) 
+              el-type (types/unnest-field (types/col-type->field return-type))]
+          {:field-type el-type 
+           :!list-expr-fn (eval `(fn [~(-> expr/args-sym (expr/with-tag RelationReader))]
```

Comment:
naming: I'd call this `:->list-expr` - `!` prefix is for things that need to be deref'd

Reply:
I'd also make `schema` available here, as it is in `emit-projection` - it's _possible_ (although unlikely) that the expression here will require it